Mathlib/CategoryTheory/Sites/Coherent/RegularSheaves.lean

```diff
@@ -26,65 +27,262 @@ namespace CategoryTheory
+
+theorem EqualizerCondition.mk (P : Cᵒᵖ ⥤ Type*)
+    (hP : ∀ (X B : C) (π : X ⟶ B) [EffectiveEpi π] [HasPullback π π], Function.Bijective
+    (MapToEqualizer P π (pullback.fst (f := π) (g := π)) (pullback.snd (f := π) (g := π))
+    pullback.condition)) : EqualizerCondition P := by
+  intro X B π _ c hc
+  have : HasPullback π π := ⟨c, hc⟩
+  specialize hP X B π
+  rw [Types.type_equalizer_iff_unique]
+  rw [Function.bijective_iff_existsUnique] at hP
+  intro b hb
+  have h₁ : ((pullbackIsPullback π π).conePointUniqueUpToIso hc).hom ≫ c.fst =
+    pullback.fst (f := π) (g := π) := by simp
+  have hb' : P.map (pullback.fst (f := π) (g := π)).op b = P.map pullback.snd.op b := by
+    rw [← h₁, op_comp, FunctorToTypes.map_comp_apply, hb]
+    simp [← FunctorToTypes.map_comp_apply, ← op_comp]
+  obtain ⟨a, ha₁, ha₂⟩ := hP ⟨b, hb'⟩
+  refine ⟨a, ?_, ?_⟩
+  · simpa [MapToEqualizer] using ha₁
+  · simpa [MapToEqualizer] using ha₂
+
+lemma equalizerCondition_iff_isIso_lift_w (P : Cᵒᵖ ⥤ Type*) {X B : C} (π : X ⟶ B)
+    [HasPullback π π] : P.map π.op ≫ P.map (pullback.fst (f := π) (g := π)).op =
+    P.map π.op ≫ P.map (pullback.snd).op := by
+  simp only [← Functor.map_comp, ← op_comp, pullback.condition]
+
+lemma mapToEqualizer_eq_comp (P : Cᵒᵖ ⥤ Type*) {X B : C} (π : X ⟶ B)
+    [HasPullback π π] : MapToEqualizer P π pullback.fst pullback.snd pullback.condition =
+    equalizer.lift (P.map π.op) (equalizerCondition_iff_isIso_lift_w P π) ≫
+    (Types.equalizerIso _ _).hom := by
+  rw [← Iso.comp_inv_eq (α := Types.equalizerIso _ _)]
+  apply equalizer.hom_ext
+  aesop
+
+/-- An alternative phrasing of the explicit equalizer condition, using more categorical language. -/
+theorem equalizerCondition_iff_isIso_lift (P : Cᵒᵖ ⥤ Type*) : EqualizerCondition P ↔
+    ∀ (X B : C) (π : X ⟶ B) [EffectiveEpi π] [HasPullback π π], IsIso
+    (equalizer.lift (P.map π.op) (equalizerCondition_iff_isIso_lift_w P π)) := by
+  constructor
+  · intro hP X B π _ _
+    have h := hP.bijective_mapToEqualizer_pullback _ X B π
+    rw [← isIso_iff_bijective, mapToEqualizer_eq_comp] at h
+    exact IsIso.of_isIso_comp_right (equalizer.lift (P.map π.op)
+      (equalizerCondition_iff_isIso_lift_w P π))
+      (Types.equalizerIso _ _).hom
+  · intro hP
+    apply EqualizerCondition.mk
+    intro X B π _ _
+    rw [mapToEqualizer_eq_comp, ← isIso_iff_bijective]
+    infer_instance
+
+/-- `P` satisfies the equalizer condition iff its precomposition by an equivalence does. -/
+theorem equalizerCondition_iff_of_equivalence (P : Cᵒᵖ ⥤ D)
+    (e : C ≌ E) : EqualizerCondition P ↔ EqualizerCondition (e.op.inverse ⋙ P) :=
+  ⟨fun h ↦ equalizerCondition_precomp_of_preservesPullback P e.inverse h, fun h ↦
+    equalizerCondition_of_natIso (e.op.funInvIdAssoc P)
+      (equalizerCondition_precomp_of_preservesPullback (e.op.inverse ⋙ P) e.functor h)⟩
+
+open WalkingParallelPair WalkingParallelPairHom in
+theorem parallelPair_initial_of_zigzag_zero {X Y : C} (f g : X ⟶ Y)
```

Comment:
I would suggest the following lemmas which could go into a new file `CategoryTheory.Limits.Final.ParallelPair`:
```lean
lemma _root_.CategoryTheory.Limits.parallelPair_initial_mk' {X Y : C} (f g : X ⟶ Y)
    (h₁ : ∀ Z, Nonempty (X ⟶ Z))
    (h₂ : ∀ ⦃Z : C⦄ (i j : X ⟶ Z),
      Zigzag (J := CostructuredArrow (parallelPair f g) Z)
        (CostructuredArrow.mk (Y := zero) i)
        (CostructuredArrow.mk (Y := zero) j)) : (parallelPair f g).Initial where
  out Z := by
    have : Nonempty (CostructuredArrow (parallelPair f g) Z) :=
      ⟨CostructuredArrow.mk (Y := zero) (h₁ Z).some⟩
    have : ∀ (x : CostructuredArrow (parallelPair f g) Z), Zigzag x
      (CostructuredArrow.mk (Y := zero) (h₁ Z).some) := by
        rintro ⟨(_|_), ⟨⟩, φ⟩
        · apply h₂
        · refine Relation.ReflTransGen.trans ?_ (h₂ (f ≫ φ) _)
          exact Relation.ReflTransGen.single (Or.inr ⟨CostructuredArrow.homMk left⟩)
    exact zigzag_isConnected (fun x y => (this x).trans (zigzag_symmetric (this y)))

lemma _root_.CategoryTheory.Limits.parallelPair_initial_mk {X Y : C} (f g : X ⟶ Y)
    (h₁ : ∀ Z, Nonempty (X ⟶ Z))
    (h₂ : ∀ ⦃Z : C⦄ (i j : X ⟶ Z), ∃ (a : Y ⟶ Z), i = f ≫ a ∧ j = g ≫ a) :
    (parallelPair f g).Initial :=
  parallelPair_initial_mk' f g h₁ (fun Z i j => by
    obtain ⟨a, rfl, rfl⟩ := h₂ i j
    have z₁ : Zigzag (J := CostructuredArrow (parallelPair f g) Z)
      (CostructuredArrow.mk (Y := zero) (f ≫ a))
      (CostructuredArrow.mk (Y := one) a) := Relation.ReflTransGen.single
        (Or.inl ⟨CostructuredArrow.homMk left⟩)
    have z₃ : Zigzag (J := CostructuredArrow (parallelPair f g) Z)
      (CostructuredArrow.mk (Y := one) a)
      (CostructuredArrow.mk (Y := zero) (g ≫ a)) := Relation.ReflTransGen.single
        (Or.inr ⟨CostructuredArrow.homMk right⟩)
    exact z₁.trans z₃)
```
I believe that the general API for zigzags could be improved. A first step would be to remove the `reducible` tag to `Zigzag` and `Zag`, and introduce suitable lemmas (like `Zag.of_hom`, `zag.of_inv`, `Zag.symm`, `Zigzag.trans`, etc.) because I have felt that there is too much `Relation.ReflTransGen` API leakage.